### DISCUSSION TITLE
Bug - I2S communication disrupted by arbitrary function call following I2C initialization.

### DISCUSSION THREAD

**Discussion by ajbradburn** [micropython/micropython#14009] [Hardware & Peripherals]
Bug - I2S communication disrupted by arbitrary function call following I2C initialization.

I have discovered that I2S communication is somehow disrupted by initializing I2C, and then calling any function.

A work-around is to do something as simple a call ‘print’ after the arbitrary function call.

Here is an example of a script that demonstrates the bug:

```

from machine import I2C, Pin, I2S

def emptyFunction():
    return

i2c = I2C(0, scl=3, sda=8)
ser = emptyFunction()
print('Safety Print') # CRITICALLY IMPORTANT. This print statement enables I2S to work properly. Some combination of the i2c assignment, and then a function call afterwards makes the I2S sound terrible. Simply printing something resolves the issue… Comment this print call to experience bug.

bclk = Pin(5) # Clock Pin
lrc = Pin(6) # Channel Select Pin
din = Pin(4) # Data Pin

audio_out = I2S(
    1, 
    sck=bclk, ws=lrc, sd=din, 
    mode=I2S.TX,
    bits=16,
    format=I2S.MONO, 
    rate=44100,
    ibuf=1024
    )

def playAudio(file):
    print(file)
    if True:
        wav_file = file
        wav = open(wav_file,'rb')
        pos = wav.seek(44) 

        wav_samples = bytearray(1024)
        wav_samples_mv = memoryview(wav_samples)
        audio_out.shift

        while True:
            num_read = wav.readinto(wav_samples_mv)
            if num_read == 0:
                break
            num_written = 0
            # Increase the volume.
            #I2S.shift(buf=wav_samples_mv, bits=16, shift=1)
            while num_written < num_read:
                num_written += audio_out.write(wav_samples_mv[num_written:num_read])
        wav.close()

playAudio('snake.wav')

audio_out.deinit()

```

I have tested this with blocking, and non blocking I2S implementation with the same result.

I discovered this while trying to operate a PCA9685 and a Max98357 on an ESP32-S3-WROOM-1 using Micropython v1.22.2 (2024-02-22).

Files can be found here:
https://github.com/ajbradburn/ESP32-PuppetRemote/blob/main/code/sound_test.py
https://github.com/ajbradburn/ESP32-PuppetRemote/blob/main/code/sound_test2.py
https://github.com/ajbradburn/ESP32-PuppetRemote/blob/main/code/snake.wav

**Comment by jimmo** [comment]
Thanks @ajbradburn -- converting this to an issue (#14050), we'll need to try and repro it and investigate.
